internal/querycoordv2/meta/resource_manager.go

```diff
@@ -0,0 +1,593 @@
+	return rg.nodes.Contain(id)
+}
+
+func (rg *ResourceGroup) GetNodes() []int64 {
+	return rg.nodes.Collect()
+}
+
+func (rg *ResourceGroup) GetCapacity() int {
+	return rg.capacity
+}
+
+type ResourceManager struct {
+	groups  map[string]*ResourceGroup
+	store   Store
+	nodeMgr *session.NodeManager
+
+	rwmutex sync.RWMutex
+}
+
+func NewResourceManager(store Store, nodeMgr *session.NodeManager) *ResourceManager {
+	groupMap := make(map[string]*ResourceGroup)
+	groupMap[DefaultResourceGroupName] = NewResourceGroup(1000000)
+	return &ResourceManager{
+		groups:  groupMap,
+		store:   store,
+		nodeMgr: nodeMgr,
+	}
+}
+
+func (rm *ResourceManager) AddResourceGroup(rgName string) error {
+	rm.rwmutex.Lock()
+	defer rm.rwmutex.Unlock()
+	if len(rgName) == 0 {
+		return ErrRGNameIsEmpty
+	}
+
+	if rm.groups[rgName] != nil {
+		return ErrRGAlreadyExist
+	}
+
+	rm.groups[rgName] = NewResourceGroup(0)
+	err := rm.saveResourceGroupInStore(rgName)
+	if err != nil {
+		log.Info("failed to add resource group",
+			zap.String("rgName", rgName),
+			zap.Error(err),
+		)
+		delete(rm.groups, rgName)
+		return err
+	}
+
+	log.Info("add resource group",
+		zap.String("rgName", rgName),
+	)
+	return nil
+}
+
+func (rm *ResourceManager) RemoveResourceGroup(rgName string) error {
+	rm.rwmutex.Lock()
+	defer rm.rwmutex.Unlock()
+	if rgName == DefaultResourceGroupName {
+		return ErrDeleteDefaultRG
+	}
+
+	if rm.groups[rgName] == nil {
+		// delete a non-exist rg should be tolerable
+		return nil
+	}
+
+	if len(rm.groups[rgName].GetNodes()) != 0 {
+		return ErrDeleteNonEmptyRG
+	}
+
+	err := rm.store.RemoveResourceGroup(rgName)
+	if err != nil {
+		log.Info("failed to remove resource group",
+			zap.String("rgName", rgName),
+			zap.Error(err),
+		)
+		return err
+	}
+	delete(rm.groups, rgName)
+
+	log.Info("remove resource group",
+		zap.String("rgName", rgName),
+	)
+	return nil
+}
+
+func (rm *ResourceManager) AssignNode(rgName string, node int64) error {
+	rm.rwmutex.Lock()
+	defer rm.rwmutex.Unlock()
+	return rm.assignNode(rgName, node)
+}
+
+func (rm *ResourceManager) assignNode(rgName string, node int64) error {
+	if rm.groups[rgName] == nil {
+		return ErrRGNotExist
+	}
+
+	if rm.nodeMgr.Get(node) == nil {
+		return ErrNodeNotExist
+	}
+
+	if ok, _ := rm.nodeMgr.IsStoppingNode(node); ok {
+		return ErrNodeStopped
+	}
+
+	rm.checkRGNodeStatus(rgName)
+	if rm.checkNodeAssigned(node) {
+		return ErrNodeAlreadyAssign
+	}
+
+	err := rm.groups[rgName].assignNode(node)
+	if err != nil {
+		return err
+	}
+
+	err = rm.saveResourceGroupInStore(rgName)
+	if err != nil {
+		log.Info("failed to add node to resource group",
+			zap.String("rgName", rgName),
+			zap.Int64("node", node),
+			zap.Error(err),
+		)
+		rm.groups[rgName].unassignNode(node)
+		return err
+	}
+
+	log.Info("add node to resource group",
+		zap.String("rgName", rgName),
+		zap.Int64("node", node),
+	)
+
+	return nil
+}
+
+func (rm *ResourceManager) saveResourceGroupInStore(rgName string) error {
```

Comment:
I'd better to see this method accepts a resource group as parameter, not a name, so we can pass a rg to save before we write it into memory table.